### PR TITLE
Support for Any

### PR DESCRIPTION
**Pull request by sigurdm**

### CODE REVIEW

File: lib/src/protobuf/unpacker.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of protobuf;
+
+typedef UnpackerFunction<T> = T Function(List<int> value);
+
+class Unpacker<T> {
+  final UnpackerFunction _fromBuffer;
```
Comment (nichite) [inline lib/src/protobuf/unpacker.dart]: Is a type parameter also needed here?

**Comment by nichite** [comment]
We will likely need to add 'pack' and 'unpack' to the reserved names for GeneratedMessage in meta.dart as well.

File: lib/src/protobuf/unpacker.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/protobuf/unpacker.dart]: Yes!
Fixed

**Comment by sigurdm** [comment]
> We will likely need to add 'pack' and 'unpack' to the reserved names for GeneratedMessage in meta.dart as well.

I'm not sure it is needed. The methods are only injected on Any. And any only has the fields of typeUrl and value. I might be overlooking something though...

**Comment by nichite** [comment]
Ah, you're right. the reserved names should be specifically for message types--since Any.pack is static from Any, and .unpack is only when that proto field is now of type any (like you said), it should be good.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## 0.10.0
+
+* Support for [any](https://developers.google.com/protocol-buffers/docs/proto3#any) messages.
```
Comment (mkustermann) [inline CHANGELOG.md]: Maybe add
```
* Adds a new `BuilderInfo.packageName` field as well as a `BuilderInfo.fullName` getter.
```

Comment (sigurdm) [reply CHANGELOG.md]: Done

File: lib/src/protobuf/unpack.dart [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+///
+/// Throws a [InvalidProtocolBufferException] if [typeUrl] does not correspond
+/// with the type of [instance].
+///
+/// This is a helper method for `Any.unpack`.
+void unpackInto<T extends GeneratedMessage>(
+    T instance, List<int> value, String typeUrl,
+    {ExtensionRegistry extensionRegistry = ExtensionRegistry.EMPTY}) {
+  String typeName = instance.info_.fullName;
+  if (typeName != _typeNameFromUrl(typeUrl)) {
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/protobuf/unpack.dart]: Good idea. Done

File: lib/src/protobuf/unpack.dart [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    throw new InvalidProtocolBufferException.wrongAnyMessage(
+        _typeNameFromUrl(typeUrl), typeName);
+  }
+  instance.mergeFromBuffer(value, extensionRegistry);
+}
+
+String _typeNameFromUrl(String typeUrl) {
+  int index = typeUrl.lastIndexOf('/');
+  return index == -1 ? typeUrl : typeUrl.substring(index + 1);
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/src/protobuf/unpack.dart]: I am unaware of such documentation. I guess we better do what the java code does.

Fixed

**Comment by sigurdm** [comment]
Now the package name is generated as a const PackageName('foo') allowing dart2js to canonicalize these.

PTAL.

File: lib/src/protobuf/generated_message.dart [hunk 5]
```diff
@@ -345,3 +345,10 @@ abstract class GeneratedMessage {
   /// For generated code only.
   void $_setInt64(int index, Int64 value) => _fieldSet._$set(index, value);
 }
+
+/// The package name of a protobuf message.
+class PackageName {
+  final String name;
+  const PackageName(this.name);
+  String get prefix => name == '' ? '' : '$name.';
+}
```
Comment (nichite) [inline lib/src/protobuf/generated_message.dart]: Looks like the no newline is breaking Travis, this is the only (minor) thing I see

**Comment by sigurdm** [comment]
Yes - I think a thumbs up from @rakudrama  is all we need...

File: lib/src/protobuf/generated_message.dart [hunk 6]
```diff
@@ -345,3 +345,10 @@ abstract class GeneratedMessage {
   /// For generated code only.
   void $_setInt64(int index, Int64 value) => _fieldSet._$set(index, value);
 }
+
+/// The package name of a protobuf message.
+class PackageName {
```
Comment (rakudrama) [inline lib/src/protobuf/generated_message.dart]: There are existing .proto files that define a message called PackageName.
Make sure this works (e.g. by using a library prefix in generated code).

Comment (sigurdm) [reply lib/src/protobuf/generated_message.dart]: Yes, I should solve https://github.com/dart-lang/dart-protoc-plugin/issues/47 first.

Comment (nichite) [reply lib/src/protobuf/generated_message.dart]: Now that the library prefix change has been merged, is this ready to go?

Comment (sigurdm) [reply lib/src/protobuf/generated_message.dart]: Now the corresponding protoc  change has landed, and a new version is published!

Comment (nichite) [reply lib/src/protobuf/generated_message.dart]: Nice! Great work on this

**Pull request merged**